File: python_poc/utils/RND_adapter.py
# Copyright (c) TUT Tampere University of Technology 2015-2018.
# This software has been developed in Procem-project funded by Business Finland.
# This code is licensed under the MIT license.
# See the LICENSE.txt in the project root for the license terms.
#
# Main author(s): Ville Heikkila, Otto Hylli, Pekka Itavuo,
#                 Teemu Laukkarinen ja Ulla-Talvikki Virta

import socket
import threading
import socketserver
import time
import random

PROCEM_SERVER_IP = "127.0.0.1"
PROCEM_SERVER_PORT = 6666

ADAPTER_IP = "127.0.0.1"
# NOTE: Port should be unique for all adapters, would be better to let OS assign it
ADAPTER_PORT = 6667

# Python string Formatter class formatted string for IoTTicket compatible message
# to get packet string the packet, call IOTTicketPKT.format(value,timestamp)
# The role of registering the devices is left to the Procem server that relays these to IOTTicket
IOTTicketPKT = '{{"name": "RandomNumber", "path": "Procem", "v": {0:d}, "ts": {1:d}, "unit": "Num", ' + \
               '"dataType": "long", "variableNumber": 3, "confidential": false }}'


# Thanks stackoverflow for this nice little lambda function
# WARNING: python is not accurate on sleeping, but depends on OS and interperter implementation
def usleep(x):
    time.sleep(x/1000000.0)


def sendUDPmsg(ip, port, message):
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        sock.sendto(bytes(message, "utf-8"), (ip, port))
    finally:
        sock.close()


# A callback class to handle incoming UDP packets (that procem server sends)
class ReceivedUDPHandler(socketserver.BaseRequestHandler):
    def handle(self):
        data = self.request[0].strip()
        socket = self.request[1]
        print("{} wrote:".format(self.client_address[0]))
        print(data)


def startUDPserver():
    server = socketserver.ThreadingUDPServer((ADAPTER_IP, ADAPTER_PORT), ReceivedUDPHandler)
    ip, port = server.server_address

    server_thread = threading.Thread(target=server.serve_forever)
    server_thread.daemon = True
    server_thread.start()
    return server


def stopUDPserver(server):
    server.shutdown()

if __name__ == "__main__":
    # start reply channel
    # srv = startUDPserver()
    while(0 < 1):
        # TODO: read meter, send message
        # IoTTicket uses ms granularity
        print(".")
        pkt = IOTTicketPKT.format(random.randint(0, 100), int(time.mktime(time.localtime()))*1000)
        sendUDPmsg(PROCEM_SERVER_IP, PROCEM_SERVER_PORT, pkt)
        usleep(1000000)
    # stopUDPserver(srv)

File: python_poc/utils/RND_adapter.c
// Copyright (c) TUT Tampere University of Technology 2015-2018.
// This software has been developed in Procem-project funded by Business Finland.
// This code is licensed under the MIT license.
// See the LICENSE.txt in the project root for the license terms.
//
// Main author(s): Ville Heikkila, Otto Hylli, Pekka Itavuo,
//                 Teemu Laukkarinen ja Ulla-Talvikki Virta


#include <stdio.h>
#include <time.h>
#include <stdint.h>
#include <pthread.h>
#include <errno.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <unistd.h> /* for close() for socket */
#include <stdlib.h>



int send_UDP_Packet(unsigned char* ip, uint16_t port, uint8_t* msg, uint32_t msg_size){
  struct sockaddr_in sa;
  int s, bytes;

  s = socket(PF_INET, SOCK_DGRAM, IPPROTO_UDP);
  if(s == -1){
    // fail, return
    return -1;
  }

  // ipv4 address
  sa.sin_family = AF_INET;
  // target addr
  sa.sin_addr.s_addr = inet_addr(ip);
  // port
  sa.sin_port = htons(port);

  bytes = sendto(s, msg, msg_size, 0, (struct sockaddr*)&sa, sizeof(sa));

  if( bytes < 0 || bytes != msg_size){
    return -1;
  } else {
    return 0;
  }
}

const unsigned char* IOTTicketPKT = "{\"name\": \"RandomNumber\", \"path\": \"ProCem/Core\", \"v\": %d, \"ts\": %d, \"unit\": \"Num\"}";

int main(){
  while(0 < 1){
    uint8_t* buf = malloc(sizeof(IOTTicketPKT)+128);
    uint32_t pkt_size = snprintf(buf, sizeof(IOTTicketPKT)+128, IOTTicketPKT, rand(), (unsigned int)time(NULL));
    if(pkt_size < sizeof(IOTTicketPKT)+128 ){
      printf(buf); printf("\n\r"); // debug
      send_UDP_Packet("127.0.0.1", 6666, buf, pkt_size);
    } else {
      perror("Run out of space in buffer\n\r");
    }
    free(buf);
    usleep(1000000);
  }
}
